public/app/core/utils/EventsProvider.tsx

```diff
@@ -0,0 +1,101 @@
+import React, { ComponentType, FunctionComponent, useEffect } from 'react';
+import { Observable, Subject, Unsubscribable } from 'rxjs';
+import { filter } from 'rxjs/operators';
+import { AppEvent, Subtract } from '@grafana/data';
+
+import appEvents from '../app_events';
+
+export interface EventsContextApi<SubscribeEvents, PublishEvents> {
+  events: (originFilter: string) => Observable<AppEvent<SubscribeEvents>>;
+  publish: (event: AppEvent<PublishEvents>, origin: string, payload?: PublishEvents) => void;
+}
+
+export interface RootEventsContextApi {
+  rootEvents: Observable<AppEvent<any>>;
+  rootPublish: (event: AppEvent<any>, payload?: any) => void;
+}
+
+const events = new Subject<AppEvent<any>>();
+const publish = <T extends {} = {}>(event: AppEvent<T>, payload?: T) => {
+  events.next({ ...event, payload });
+  appEvents.emit(event, payload);
+};
+
+export const rootEventing: RootEventsContextApi = {
+  rootEvents: events,
+  rootPublish: publish,
+};
+
+export const EventsContext = React.createContext(rootEventing);
+EventsContext.displayName = 'EventsContext';
+
+export const EventsProvider = ({ children }: { children: React.ReactNode }) => {
+  return <EventsContext.Provider value={rootEventing}>{children}</EventsContext.Provider>;
+};
+
+export const provideEvents = (component: ComponentType<any>) => (props: any) => (
+  <EventsProvider>{React.createElement(component, { ...props })}</EventsProvider>
+);
+
+export const withEvents = <SubscribeEvents extends {}, PublishEvents extends {}>(
+  ...subscribeEvents: Array<AppEvent<SubscribeEvents>>
+) => {
+  type ComponentWithEventingType = EventsContextApi<SubscribeEvents, PublishEvents>;
+  return <C extends ComponentWithEventingType>(Component: ComponentType<C>) => {
+    type ComponentWithOutEventingType = Subtract<C, ComponentWithEventingType>;
+
+    const ComponentWithEventing: FunctionComponent<ComponentWithOutEventingType> = (props: any) => {
+      useEffect(() => {
+        return function unsubscribe() {
```

Comment:
Why do we have different lifecycle for unsubscribing and subscribing? Seems like right now we will subscribe on every render while unsubscribe only on unmount. Won't that leak the subscriptions?

Reply:
That could be the case yes, but as we only subscribe in `componentDidMount` I think it's ok. But I agree that it's not solid.

Reply:
> but as we only subscribe in componentDidMount

Ah ok missed that `subscribe` later on is a function. I am not sure I like the premise that even though component does the subscription, HOC does the unsubscribe. As you say this will require component to remember to only subscribe once in onMount otherwise it could probably fail or leak which seems like an encapsulation problem with the abstraction.